src/lib/support/SparseArray.h

```diff
@@ -0,0 +1,133 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Defines a data structure that can store an array of uint64_t values
+ *      in not contiguous indexes. The array is initialized with a constant
+ *      capacity. The data structure supports serialize/deserialize operations,
+ *      where, serialize converts the array to a base64 string, and deserializing
+ *      the base64 string reconstructs the original array.
+ *
+ *      The data is stored in LittleEndian byte order in the array, so that
+ *      serialized data can be deserialized correctly on different machine
+ *      architectures.
+ *
+ */
+
+#pragma once
+
+#include <core/CHIPEncoding.h>
+#include <support/Base64.h>
+#include <support/CodeUtils.h>
+
+namespace chip {
+
+class SparseU64ArrayBase
+{
+
+public:
+    SparseU64ArrayBase(uint64_t * data, uint16_t capacity) : mData(data), mCapacity(capacity), mNextAvailable(0) {}
+
+    /**
+     * @brief
+     *   Serialize the sparse array into a base 64 encoded string.
+     *   Only the values till mNextAvailable index are encoded.
+     *   The empty indexes between 0, and mNextAvailable, are also
+     *   encoded.
+     *
+     * @param[in] buf Buffer where serialized string is written
+     * @param[in] buflen Length of buf
+     * @return pointer to buf, or nullptr in case of error
+     */
+    const char * SerializeBase64(char * buf, uint16_t & buflen);
+
+    /**
+     * @brief
+     *   Deserialize a base64 encoded string into the sparse array.
+     *   The mNextAvailable index is calculated based on how many
+     *   values are in the deserialized array.
+     *
+     * @param[in] serialized Serialized buffer
+     * @param[in] buflen Length of buffer
+     * @return CHIP_NO_ERROR in case of success, or the error code
+     */
+    CHIP_ERROR DeserializeBase64(const char * serialized, uint16_t buflen);
+
+    /**
+     * @brief
+     *   Get the length of string if the array is serialized.
+     */
+    uint16_t SerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * mNextAvailable)); }
+
+    /**
+     * @brief
+     *   Get the maximum length of string if the array was full and serialized.
+     */
+    uint16_t MaxSerializedSize() { return static_cast<uint16_t>(BASE64_ENCODED_LEN(sizeof(uint64_t) * mCapacity)); }
+
+    /**
+     * @brief
+     *   Find a value in the array.
+     *
+     * @param[in] value Value to find
+     * @return index of the value if found, or max length (mCapacity) of the array
+     */
+    uint16_t Find(uint64_t value);
```

Comment:
Hmm.  How is this usable as a public API?  What would a consumer do with the returned index?

I see this is used in Remove and in tests, but for actual API consumers a `Contains` or `Has` or something returning boolean would be more useful, I'd think.  I'm not sure whether we're willing to mark tests as `friend` or restrict them to testing API, not implementation details, so might need to keep this public, I guess.   That said, I don't see any tests testing for the precise value returned by `Find`; just testing whether it returned the capacity or not, so the boolean-returning API would be ok for the tests we have right now...